src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90

```diff
@@ -1548,7 +1548,461 @@ subroutine read_obs_sm_ASCAT(                                  &
+       
+       call read_obs_SMAP_fnames( date_time_up, this_obs_param,                &
+            fname_of_fname_list, N_fnames_max,                                 &
+            N_fnames_tmp, fname_list((N_fnames+1):(N_fnames+N_fnames_max)) )
+       
+       N_fnames = N_fnames + N_fnames_tmp
+       
+    end if
+    
+    tmpfnames = fname_list(1:N_fnames)
+    
+    ! ----------------------------------------------------------------
+    !
+    ! find files that have obs within assimilation window
+    
+    N_tmp = 0   
+    
+    do kk = 1,N_fnames
+
+       tmpfname = fname_list(kk)
+
+       ! Are we in the required assimilation window?
+       !
+       ! e.g. Y2019/M07/M01-ASCA-ASCSMO02-NA-5.0-20190702075700.000000000Z-20190702084627-1350204.bfr
+       !
+       !      12345678901234567890123456789012345678901234567890123456789012345678901234567890 
+       !               1         2         3         4         5         6         7
+
+       str_date_time = tmpfname(40:53)
+      
+       read(str_date_time( 1: 4), *) date_time_tmp%year
+       read(str_date_time( 5: 6), *) date_time_tmp%month
+       read(str_date_time( 7: 8), *) date_time_tmp%day
+       read(str_date_time( 9:10), *) date_time_tmp%hour
+       read(str_date_time(11:12), *) date_time_tmp%min
+       read(str_date_time(13:14), *) date_time_tmp%sec
+       
+       if ( datetime_lt_refdatetime( date_time_low_fname, date_time_tmp ) .and.          &
+            datetime_le_refdatetime( date_time_tmp,       date_time_up  )       ) then 
+
+          N_tmp = N_tmp + 1
+          
+          ! Remove the '/D03/' from the directory part as using "read_obs_SMAP_fnames" 
+          
+          ind = index(tmpfname, "/D")
+          
+          if (ind > 0) then
+             tmpfname2 = tmpfname(1:ind) // tmpfname(ind+5:)
+          end if
+
+          tmpfnames(N_tmp) = trim(this_obs_param%path) // '/' // trim(tmpfname2)
+          
+       end if
+       
+    end do
+    
+    fnames  = tmpfnames(1:N_tmp)
+    N_files = N_tmp
+    
+    ! ----------------------------------------------------------------
+    !
+    ! loop through files and read obs + metadata into tmp_data
+    
+    if (N_files>0) then
+
+       ! read and process data if files are found
+
+       allocate(tmp1_lon(  max_rec   ))
+       allocate(tmp1_lat(  max_rec   ))
+       allocate(tmp1_obs(  max_rec   ))
+       allocate(tmp1_jtime(max_rec   ))
+       
+       allocate(tmp_data(  max_obs,15))
+       
+       N_obs = 0
+       
+       do kk = 1,N_files
+          
+          ! open bufr file
+          
+          call closbf(lnbufr)  ! if a file with unit number lnbufr is open in (or "linked" with) BUFR, close it
+          open(lnbufr, file=trim(fnames(kk)), action='read', form='unformatted')
+          call openbf(lnbufr, 'SEC3', lnbufr) 
+          call mtinfo( trim(this_obs_param%path) // '/BUFR_mastertable/', lnbufr+1, lnbufr+2)
+          call datelen(10)
```

Comment:
A couple of comments/questions here:
1. I looked up the "openbf()" interface for [NCEPLIBS-bufr 12.0.1](https://noaa-emc.github.io/NCEPLIBS-bufr/openbf_8f.html).  The doc there says that the unitnumbers passed into openbf() must already be associated with fortran-opened files, which I don't think they are.  I suspect the library we're using now Fortran-opens the file at the same time as it is being linked with the BUFR lib, and the Fortran-open call is no longer needed, but I'm only guessing.  However, if that's the case, why do we Fortran-close the file below, as in `close(lnbufr)` ?
2. Why do we use 'SEC3' mode, which "use[s] Section 3 of input messages for decoding rather than DX BUFR table information from LUNDX" (per the doc above).  Don't we have the BUFR tables?  Since we loop through each individual record here, asking the library to figure out the BUFR table on each read may add a lot of overhead.  But here, too, I'm only guessing.

Reply:
Please disregard item 1.  Apologies, I somehow missed the Fortran open() just before the openbf().  After discussion, we can also disregard item 2.